File: api/broker/nats.js
const { connect, StringCodec } = require("nats");
const { createValidateFunction } = require("../config/validation");

const sc = StringCodec();

class ServiceBroker {
  constructor(config = {}) {
    this.debug = config.debug;
    this.servers = config.servers;
    this.token = config.token;
  }

  async start() {
    try {
      const conn = await connect({
        debug: this.debug,
        servers: this.servers,
        token: this.token,
      });
      this.broker = conn;
      if (this.broker.isClosed()) {
        throw new Error(`Broker is closed`);
      }
      return this;
    } catch (error) {
      throw error;
    }
  }

  createService(name, wokers = 1, actionSchemas = []) {
    this._isConnected();

    actionSchemas.forEach(async (actionSchema) => {
      for (let i = 1; i <= wokers; i++) {
        const topic = this._prefixTopic(
          name,
          actionSchema.version,
          actionSchema.name
        );
        this.broker.subscribe(topic, {
          queue: `${topic}-${i}`,
          callback: this._prefixActions(actionSchema),
        });
        console.log(`[${topic}]: Registered topic`);
      }
    });
  }

  emit() {
    const ctx = this;
    return async (topic, payload) => {
      try {
        ctx._isConnected();
        const m = await ctx.broker.request(topic, this._encode(payload));
        const response = this._decode(m.data);
        if (!response.ok) {
          throw new Error(response.message);
        }
        return response;
      } catch (error) {
        throw error;
      }
    };
  }

  /**
   * @returns
   */
  _prefixActions(actionSchema) {
    return async (err, msg) => {
      try {
        if (err) {
          console.error(
            `Action ${msg} exited because of error: ${err.message}`
          );
          return;
        }
        const [broker, service, actionName] = msg.subject.split(".");
        if (actionName === actionSchema.name) {
          let payload = this._decode(msg.data)
          // validate action input
          await createValidateFunction(actionSchema.validate)(payload)
          const ctx = this._context();
          ctx.payload = payload;
          
          const result = await actionSchema.handle(ctx);
          await msg.respond(
            this._encode({
              ok: true,
              result,
            })
          );
        }
      } catch (error) {
        await msg.respond(
          this._encode({
            ok: true,
            message: error.message,
          })
        );
      }
    };
  }

  _prefixTopic(name, version, schema) {
    return `v${Number(version) || "1"}.${name}.${schema}`;
  }

  _isConnected() {
    if (!this.broker) {
      throw new Error(`Broker is closed`);
    }
    return true;
  }

  _context() {
    return {
      broker: this.broker,
      emit: this.emit(this),
    };
  }

  _encode(payload) {
    return sc.encode(Buffer.from(JSON.stringify(payload)));
  }

  _decode(payload) {
    return JSON.parse(Buffer.from(payload).toString("utf-8"));
  }
}

module.exports = ServiceBroker;

File: user/broker/nats.py
import asyncio
import json
from typing import List
import nats

from common.actions import ActionSchema

# from nats.errors import ConnectionClosedError, TimeoutError, NoServersError

class ServiceBroker:
    def __init__(self, servers, token = None):
        self.servers = servers
        self.token = token

    async def start(self):
        try:
            conn = await nats.connect(**{ "servers":self.servers, "token": self.token })
            self.broker = conn
            if self.broker.is_closed:
                raise Exception('Broker is closed')
        except Exception as e:
            raise e

    async def create_service(self, name, workers=1, action_schemas: List[ActionSchema] = []):
        self._is_connected()
        for action in action_schemas:
            for i in range(workers):
                topic_name = self._prefix_topic(
                    name=name,
                    version=action.version,
                    action_name=action.name
                )
                await self.broker.subscribe(
                    subject=topic_name,
                    queue= "{ftopic}={fnum}".format(ftopic=topic_name, fnum=i),
                    cb=self._prefix_actions(action),
                )
                print("[{ftopic}]: Registered topic".format(ftopic=topic_name))

    def _prefix_actions(self, action):
        async def msg_handle(msg):
            try:
                subject = msg.subject
                reply = msg.reply
                data = msg.data.decode()
                action_name = subject.split(".")[2]
                if action_name != action.name:
                    return None

                ctx = self._context()
                ctx['payload'] = self._decode(data)
                result = await action.handle(ctx)
                await msg.respond(self._encode({"ok": True, "result": result }))
            except Exception as e:
                await msg.respond(self._encode({"ok": False, "message": str(e) }))
        return msg_handle

    def emit(self):
        ctx = self
        async def emit_handle(topic, payload):
            try:
                ctx._is_connected()
                m = await ctx.broker.request(
                    subject=topic,
                    payload=ctx._encode(payload=payload)
                )
                response = self._decode(m.data)
                if response['ok'] == False:
                    raise Exception(response['message'])
                return response
            except Exception as e:
                raise e
        return emit_handle

    def _prefix_topic(self, name, version, action_name):
        return "v{fversion}.{fname}.{faction_name}".format(
            fversion=version,
            fname=name,
            faction_name=action_name,
        )

    def _is_connected(self):
        if self.broker == None:
            raise Exception('Broker is closed')
        return True

    def _encode(self, payload):
        return json.dumps(payload).encode()
        
    def _decode(self, payload):
        return json.loads(payload)

    def _context(self):
        return {
            "broker": self.broker,
            "emit": self.emit(),
        }